RFC/src/RFC-0310_AssetImplementation.md

```diff
@@ -0,0 +1,607 @@
+    // Optionally get a set of specific unique_ids
+    repeated bytes unique_ids = 2;
+  }
+
+  message GetTokensResponse {
+    bytes unique_id = 1;
+    bytes asset_public_key = 2;
+    bytes owner_commitment = 3;
+    bytes mined_in_block = 4;
+    uint64 mined_height = 5;
+  }
+
+
+```
+
+## Validator Node Instructions
+
+> This section is specific to the Tari Validator Node. Other sidechains may make use of 
+> the base layer mechanisms for sidechains as well
+
+Callers may issue instructions to a validator node committee member using GRPC. Instructions take the form:
+
+```protobuf
+
+message Instruction{
+    // The public key of the asset being invoked
+    bytes asset_public_key =1;
+    // The template being invoked
+    uint32 template_id = 2;
+    // The method on the template being invoked.
+    string method =3;
+    // The position based arguments
+    repeated bytes args = 4;
+    // A commitment that exists on the base layer that identifies
+    // the caller. For example, the asset issuer could use the 
+    // commitment that holds the asset registration, while the owner
+    // could use the commitment on the UTXO containing the NFT
+    bytes caller_commitment = 4;
+    // A unique number to identify this instruction
+    uint64 id = 5;
+    // A comsig proving the caller owns the commitment in `caller_commitment`
+    // The challenge is the hash of the above fields, in order
+    bytes caller_sig = 6;
+    
+    // Fee fields TBD
+}
+
+```
+
+Upon receiving an instruction, a validator node must store it in its mempool and broadcast it to all other members of
+the committee. 
+
+As per Hotstuff and [RFC-0340], when a validator node becomes the leader, it selects a list of instructions from its 
+mempool and creates a proposal. 
+
+> The leader can use any algorithm to select instructions from the mempool and it is 
+valid for it to select no instructions
+
+The proposal must conform to the below protobuf
+
+```protobuf
+
+message HotStuffTreeNode {
+    // The parent node, as per Hotstuff Specification
+    bytes parent = 1;
+    
+    // The payload
+    InstructionSet payload = 2;
+}
+
+message InstructionSet{
+   // A list of instructions
+   repeated Instruction instructions = 1;
+   // The hash of the block at the tip of the base layer, according to the leader's view of the base layer
+   bytes base_layer_block = 3;
+}
+
+```
+
+> Note that the base layer tip used by the leader may be a number of blocks below the actual tip to minimise the 
+> number of reorgs and increase the chance that all committee members have the same view of the base layer
+
+The committee members process the proposal via Hotstuff, except that during the prepare and pre-commit stages, they must
+confirm that block in `base_layer_block` is still present in the base node that they are connected to. 
+
+If a validator node has reached the Commit stage, it must prevent the base layer from reorging by issuing a GRPC call to 
+`LockBlock`
```

Comment:
We need to properly think how this is implemented. This does open up a lot of attack vectors for a call like this. 

Should we not handle this the same way we handle a double spend or normal transaction for that matter. It's the validator node that cares that the checkpoint is in the main chain. A specific baseNode won't care about "Jhon's private checkpoint". If it gets reorged out they need to re-add it. Keeping old forked orphan chains breaks a lot of the block_chain ideas and needs special rules to cover those.

Reply:
Lets take this example:
There are 2 NFTs, `A` and `B`. 
At block height 1 Pete has 1 utxo with funds. 
At block height 2 Pete buys a NFT `A` with those funds. 
At block height 3 we have a checkpoint confirming Pete has NFT `A`

Now the block chain reorgs back to height 1
At block height 2-1 Pete buys a NFT `B` with those funds. 
At block height 3-1 we have a checkpoint confirming Pete has NFT `B`

So now at block height 3-1 what does Pete own? NFT `A `or `B` because according to the "LockedBlock" he owns `A`, but according to the main chain he owns `B`. Or does he own both and he just claimed a free NFT? When you want to buy NFT `A`, who owns it?

According to blockchain rules, NFT `A` is the person who owned `A` at height 1. As in the history Pete never bought `A`

Reply:
Let me expand a bit on the rational of it. The problem it is trying to solve is the fact that all committee members must have the same view of the blockchain when they are deciding if an instruction is valid or not. You could remove this probability by requiring commitments to be confirmed by X blocks before they can be used as a proof, but once a Hotstuff node commits to executing a set of instructions, it can't have the landscape underneath it change. It has already told the rest of the committee it will execute the instructions, so there has to be some sort of concept of locking.

Perhaps one way of doing it is to validate and pre-execute all the instructions in the Commit or PreCommit states and then the Decide phase is simply updating the state. I do think that the block that is at the tip should be communicated to the other committee nodes, so that they can confirm their view of the chain is the same